engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollingformula/BaseRollingFormulaOperator.java

```diff
@@ -0,0 +1,291 @@
+            @NotNull final MatchPair pair,
+            @NotNull final String[] affectingColumns,
+            @Nullable final String timestampColumnName,
+            final long reverseWindowScaleUnits,
+            final long forwardWindowScaleUnits,
+            @NotNull final String formula,
+            @NotNull final String paramToken,
+            @NotNull final Map<Class<?>, FormulaColumn> formulaColumnMap,
+            @NotNull final TableDefinition tableDef) {
+        super(pair, affectingColumns, timestampColumnName, reverseWindowScaleUnits, forwardWindowScaleUnits, true);
+        this.formulaColumnMap = formulaColumnMap;
+        this.tableDef = tableDef;
+
+        final String outputColumnName = pair.leftColumn;
+
+        final Class<?> columnType = tableDef.getColumn(pair.rightColumn).getDataType();
+        vectorType = getVectorType(columnType);
+
+        // Handle the rare (and probably not useful) case where the formula is an identity formula. We need to make
+        // a copy of the RingBuffer wrapper and store that as a DirectVector. If not, we will point to the live data
+        // in the ring.
+        final String formulaToUse = formula.equals(paramToken) ? formula + ".getDirect()" : formula;
+
+        // Re-use the formula column if it's already been created for this type. No need to synchronize; these
+        // operators are created serially.
+        // TODO: does generic Object need to be handled uniquely?
+        formulaColumn = formulaColumnMap.computeIfAbsent(columnType, t -> {
+            final FormulaColumn tmp = FormulaColumn.createFormulaColumn(outputColumnName,
+                    FormulaUtil.replaceFormulaTokens(formulaToUse, paramToken, PARAM_COLUMN_NAME));
+
+            final ColumnDefinition<?> inputColumnDefinition = ColumnDefinition
+                    .fromGenericType(PARAM_COLUMN_NAME, vectorType, columnType);
+            tmp.initDef(Collections.singletonMap(PARAM_COLUMN_NAME, inputColumnDefinition));
+            return tmp;
+        });
+    }
+
+    protected BaseRollingFormulaOperator(
+            @NotNull final MatchPair pair,
+            @NotNull final String[] affectingColumns,
+            @Nullable final String timestampColumnName,
+            final long reverseWindowScaleUnits,
+            final long forwardWindowScaleUnits,
+            final Class<?> vectorType,
+            @NotNull final Map<Class<?>, FormulaColumn> formulaColumnMap,
+            @NotNull final TableDefinition tableDef) {
+        super(pair, affectingColumns, timestampColumnName, reverseWindowScaleUnits, forwardWindowScaleUnits, true);
+        this.formulaColumnMap = formulaColumnMap;
+        this.tableDef = tableDef;
+
+        final Class<?> columnType = tableDef.getColumn(pair.rightColumn).getDataType();
+        this.vectorType = vectorType;
+
+        // Re-use the formula column that already been created for this type.
+        // TODO: does generic Object need to be handled uniquely?
```

Comment:
Don't forget to remove this TODO (I believe it only applies to the above constructor anyway).